src/main/java/com/rocketden/main/service/RoomService.java

```diff
@@ -298,4 +299,25 @@ public RoomDto updateRoomSettings(String roomId, UpdateSettingsRequest request)
         socketService.sendSocketUpdate(roomDto);
         return roomDto;
     }
+
+    public RoomDto setSpectator(String roomId, Boolean isSpectator, SetSpectatorRequest request) {
+        Room room = repository.findRoomByRoomId(roomId);
+
+        // Return error if room could not be found
+        if (room == null) {
+            throw new ApiException(RoomError.NOT_FOUND);
+        }
+
+        // Return error if the initiator is not the host or the same as the receiver
+        User initiator = UserMapper.toEntity(request.getInitiator());
+        User receiver = UserMapper.toEntity(request.getReceiver());
+        if (!room.getHost().equals(initiator) && !initiator.equals(receiver)) {
+            throw new ApiException(RoomError.INVALID_PERMISSIONS);
+        }
+
+        User modifiedUser = room.getUserByUserId(receiver.getUserId());
+        modifiedUser.setIsSpectator(isSpectator);
```

Comment:
You need to handle null values here:
1. The values `initiator` and `receiver` have to be verified as non-null.
2. The `modifiedUser` below can be null if the `receiver.getUserId()` is invalid, meaning that you also have to confirm that the Users provided actually exist in the database by checking against `findUserById` and verifying that you get a non-null result.